keps/sig-cli/5295-kyaml/README.md

```diff
@@ -0,0 +1,1500 @@
+
+To make this change, we will:
+1. Create new code (see details below) which serializes any object into KYAML.
+2. Introduce a kubectl output format, `-o=kyaml`, which uses this new package.
+3. Create tooling which reformats YAML or JSON as KYAML (see details below).
+4. Work with SIG Docs to update examples to this format, both example command-lines and examples of API objects.
+5. Look for other examples across kubernetes-affiliated git repositories, and offer PRs to convert to KYAML.
+6. Make a lot of noise about it, and try to get more people to use it
+
+The rationale for changing everything everywhere is simple.
+Many users start with examples,
+which they copy and modify.
+By changing the examples we seed the change in the ecosystem.
+One success metric might be
+that KYAML becomes the preferred format in the most common helm charts.
+
+### Notes/Constraints/Caveats
+
+**Important note:**
+This is EXCLUSIVELY about client-side formatting.
+This KEP *does not* propose any changes to apiserver
+or other components besides kubectl.
+
+Syntax has always been a magnet for arguments in our industry.
+We expect this to be no exception.
+Reasonable people will disagree with our choices,
+which is fine, but we should hold a strong opinion on this.
+
+YAML is a very large and complex specification.
+We expect that we will find places where KYAML may be less readable than conventional YAML.
+We should revisit this specification when that happens.
+Given that KYAML is YAML it should be possible to revise
+without breaking anyone.
+
+### Risks and Mitigations
+
+Risk: Our encoder produces some output which is not valid YAML.
+Mitigation: That’s a bug. It’s hard to imagine a case that can’t be fixed.
+
+Risk: The result could be net worse than plain YAML.
+Mitigation: We will produce many examples and include them in the review process. We will solicit feedback from various groups on the result.
+
+Risk: This gives a false sense of safety, but doesn’t SOLVE  the problems with YAML, just avoids them.
+Mitigation: Docs?
+
+Risk: Like all APIs, people will come to depend on the details, which may prevent us from changing them.
+Mitigation: Be as precise as possible where we are confident, but leave room where we are not.
+
+## Design Details
+
+Creating an encoder for this format is not very technically challenging,
+but there are many details to consider.
+
+### Feature gate
+
+Literal feature gates are not super useful for CLI features, so SIG CLI has a
+similar mechanism to gate features via environment variables.  We will use a
+variable named `"KUBECTL_KYAML"` to gate this feature.
+
+The output flags for JSON and YAML are handled in
+`k8s.io/cli-runtime/pkg/genericclioptions`.  Unfortunately, that package is
+already transitively depended on by `k8s.io/kubectl/pkg/cmd/util` where the CLI
+gates are defined. Rather than trying to break the import cycle, we will just
+check the environment variable directly, without the utils package.
```

Comment:
This is reasonable for now, even if you just copy&paste the similar functions. In the long run I'm planning to rewrite the mechanism we currently have and ideally re-use the one which is part of component-base package.

Reply:
ACK.  It's simple enough :)